src/qtviewer/IfcViewerWidget.cpp

```diff
@@ -1,27 +1,170 @@
 }
 
 void IfcViewerWidget::paintGL()
 {
-    // Render geometries from the parsed IFC file
-    // Draw the scene:
-    QOpenGLFunctions *f = QOpenGLContext::currentContext()->functions();
-    f->glClear(GL_COLOR_BUFFER_BIT); 
+    // Render OSG scene
+    m_viewer->frame();
+}
+
+void IfcViewerWidget::mouseMoveEvent(QMouseEvent *event)
+{
+    this->getEventQueue()->mouseMotion(event->position().x() * m_dpiScale, event->position().y() * m_dpiScale);
+}
+
+void IfcViewerWidget::mousePressEvent(QMouseEvent *event)
+{
+    unsigned int button = this->getMouseButtonNum(event);
+    this->getEventQueue()->mouseButtonPress(event->position().x() * m_dpiScale, event->position().y() * m_dpiScale, button);
+}
+
+void IfcViewerWidget::mouseReleaseEvent(QMouseEvent *event)
+{
+    unsigned int button = this->getMouseButtonNum(event);
+    this->getEventQueue()->mouseButtonRelease(event->position().x() * m_dpiScale, event->position().y() * m_dpiScale, button);
+}
+
+void IfcViewerWidget::wheelEvent(QWheelEvent *event)
+{
+    int delta = event->angleDelta().y();
+    osgGA::GUIEventAdapter::ScrollingMotion motion = delta > 0 ?
+            osgGA::GUIEventAdapter::SCROLL_UP : osgGA::GUIEventAdapter::SCROLL_DOWN;
+    this->getEventQueue()->mouseScroll(motion);
+}
+
+bool IfcViewerWidget::event(QEvent* event)
+{
+    bool handled = QOpenGLWidget::event(event);
+    this->update();
+    return handled;
+}
+
+osgGA::EventQueue* IfcViewerWidget::getEventQueue() const 
+{
+    osgGA::EventQueue* eventQueue = m_graphicsWindow->getEventQueue();
+    return eventQueue;
+}
+
+unsigned int IfcViewerWidget::getMouseButtonNum(QMouseEvent* event)
+{
+    unsigned int button = 0;
+    switch (event->button()){
+    case Qt::LeftButton:
+        button = 1;
+        break;
+    case Qt::MiddleButton:
+        button = 2;
+        break;
+    case Qt::RightButton:
+        button = 3;
+        break;
+    default:
+        break;
+    }
+    return button;
+}
+
+void IfcViewerWidget::loadFile(const std::string& filePath)
+{
+    std::vector<osg::ref_ptr<osg::Geometry>> geometries;
+
+    if(!m_parser.Parse(filePath, geometries)) {
+        MessageLogger::log("Failed to prepare IFC geometry");
+        return;
+    }
+
+    MessageLogger::log("\nIFC geometry prepared");
+
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode;
+    for (auto geometry : geometries) {
+        geode->addDrawable(geometry);
+    }
+
+    prepareSceneWithGeometry(geode);
+
+    m_viewer->setCameraManipulator(m_mouseHandler);
+}
+
+void IfcViewerWidget::prepareSceneWithGeometry(osg::ref_ptr<osg::Geode> geode)
+{
+    // Apply a rotation to the model
+    // MatrixTransform allows applying transformations (translation, rotation, scaling) to its children
+    osg::ref_ptr<osg::MatrixTransform> rotationTransform = new osg::MatrixTransform;
+    rotationTransform->setMatrix(osg::Matrix::rotate(osg::DegreesToRadians(45.0), osg::Vec3(0.0, 0.0, 1.0)));
+    rotationTransform->addChild(geode);
```

Comment:
I would propose to solve this in the camera orientation as opposed to rotating the model. For example, if we later add a X,Y,Z axis indicator it wouldn't be in sync with the model when rotated. Maybe use something like `ComputeBoundsVisitor` to get a bounding box of the scene and `viewer->getCamera()->setViewMatrixAsLookAt()`. Take the bounding box center as the lookat target, center + radius * some_vector as the lookat eye position.

Reply:
Absolute agree. Removed rotation transformation and applied camera orientation upon loading the IFC.